### PR TITLE
[BugFix] Should always get db lock when analyze

### PR DESCRIPTION
**Pull request by kangkaisen**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

```
2023-04-20 11:17:03,394 WARN (starrocks-mysql-nio-pool-24|751) [StmtExecutor.execute():558] execute Exception, sql select sum(cast(v1 as int)), sum(v2), sum(v3), max(v4), max(v5), min(v6), min(v7), sum(v8) from primary_table_2000_abnormal where k1 > '2020-08-02'
java.util.ConcurrentModificationException: null
        at java.util.TreeMap$PrivateEntryIterator.nextEntry(TreeMap.java:1211) ~[?:1.8.0_322]
        at java.util.TreeMap$EntryIterator.next(TreeMap.java:1247) ~[?:1.8.0_322]
        at java.util.TreeMap$EntryIterator.next(TreeMap.java:1242) ~[?:1.8.0_322]
        at java.util.HashMap.putMapEntries(HashMap.java:513) ~[?:1.8.0_322]
        at java.util.HashMap.<init>(HashMap.java:491) ~[?:1.8.0_322]
        at com.google.common.collect.Maps.newHashMap(Maps.java:244) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.catalog.OlapTable.copyOnlyForQuery(OlapTable.java:290) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.analyzer.AnalyzerUtils$OlapTableCollector.visitTable(AnalyzerUtils.java:497) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.analyzer.AnalyzerUtils$OlapTableCollector.visitTable(AnalyzerUtils.java:484) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.ast.TableRelation.accept(TableRelation.java:151) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.ast.AstVisitor.visit(AstVisitor.java:55) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.ast.AstVisitor.visit(AstVisitor.java:51) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.ast.AstTraverser.visitSelect(AstTraverser.java:90) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.ast.SelectRelation.accept(SelectRelation.java:221) ~[starrocks-fe.jar:?]
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 16 (81.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|13|16|81.25%|[199, 200, 202]|

**Review by liuyehcf — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java [hunk 1]
```diff
@@ -179,11 +182,26 @@ public static ExecPlan createQueryPlanWithReTry(QueryStatement queryStmt,
...
+                unLock(dbs);
+            }
 
             // Only need to re analyze and re transform when schema isn't valid
             if (i > 0 && !isSchemaValid) {
-                Analyzer.analyze(queryStmt, session);
+
+                try {
+                    // We always need db lock when analyze phase
+                    lock(dbs);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:199]: What happens If database has been modified?

Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:197]: fixed

**Review by packy92 — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 15 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|12|15|80.00%|[197, 198, 200]|

File: fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java [hunk 2]
```diff
@@ -179,11 +180,26 @@ public static ExecPlan createQueryPlanWithReTry(QueryStatement queryStmt,
             long planStartTime = System.currentTimeMillis();
 
             Set<OlapTable> olapTables = Sets.newHashSet();
-            AnalyzerUtils.copyOlapTable(queryStmt, olapTables);
+            Map<String, Database> dbs = AnalyzerUtils.collectAllDatabase(session, queryStmt);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:183]: also need reset `session.setCurrentSqlDbIds(dbs.values().stream().map(Database::getId).collect(Collectors.toSet()));` ?

**Review by packy92 — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 16 (81.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|13|16|81.25%|[199, 200, 202]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**